### PR TITLE
Cmake: fix compilation of assembly files with MSVC builds

### PR DESCRIPTION
**Pull request by zefklop**
## Purpose

Improve situation regarding the way we build assembly files, in order to fix build within VIsual Studio.

JIRA issue: [CORE-17423](https://jira.reactos.org/browse/CORE-17423)

## Proposed changes

- Deduplicate some symbols from msvcrt & msvcrtex
- Use ASM_MASM language when building with MSVC
- Get rid of add_asm_files function

### CODE REVIEW

**Comment by Extravert-ir** [comment]
Does `ASM_MASM` work with `MSVC_IDE` build?

**Comment by zefklop** [comment]
> Does `ASM_MASM` work with `MSVC_IDE` build?

It's supposed to, but I am unable to test atm :-/

**Comment by Extravert-ir** [comment]
I don't know what's going on on my computer, but it doesn't work :(
```
R:\>cmake --build . --target ntoskrnl
Microsoft (R) Build Engine version 16.8.2+25e4d540b for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Building Custom Rule C:/rosgit/CMakeLists.txt
  Generating bugcodes.mc
  The system cannot find the path specified.
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(238,5): error MSB8066: Custom build for 'R:\CMak eFiles\6ce3a34193f0f6c091fcc615959fa617\bugcodes.mc.rule;R:\CMakeFiles\6ce3a34193f0f6c091fcc615959fa617\bugcodes.h.rule;R:\CMakeFiles\e15018c7a0ada3bc3dfb67bf253964 54\bugcodes.rule' exited with code 3. [R:\sdk\include\reactos\mc\bugcodes.vcxproj]
  Generating wdm.h
  The system cannot find the path specified.
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(238,5): error MSB8066: Custom build for 'R:\CMak eFiles\b30ab84552bfb0cfdf54f4dabcbd020b\wdm.h.rule;R:\CMakeFiles\8deb71bd27f562996635ca4c6c58a76c\wdm.h.rule;R:\CMakeFiles\b30ab84552bfb0cfdf54f4dabcbd020b\ntddk.h. rule;R:\CMakeFiles\8deb71bd27f562996635ca4c6c58a76c\ntddk.h.rule;R:\CMakeFiles\b30ab84552bfb0cfdf54f4dabcbd020b\ntifs.h.rule;R:\CMakeFiles\8deb71bd27f562996635ca4c6 c58a76c\ntifs.h.rule;R:\CMakeFiles\b30ab84552bfb0cfdf54f4dabcbd020b\devioctl.h.rule;R:\CMakeFiles\45e764fb58ff47b776f0c88022268775\devioctl.h.rule;R:\CMakeFiles\b30 ab84552bfb0cfdf54f4dabcbd020b\ntdef.h.rule;R:\CMakeFiles\45e764fb58ff47b776f0c88022268775\ntdef.h.rule;R:\CMakeFiles\b30ab84552bfb0cfdf54f4dabcbd020b\winnt.h.rule;R :\CMakeFiles\45e764fb58ff47b776f0c88022268775\winnt.h.rule;R:\CMakeFiles\58cdf411284e9d98347a40e78759a0ac\xdk.rule' exited with code 3. [R:\sdk\include\xdk\xdk.vcxp roj]

R:\>
```

**Comment by learn-more** [comment]
> cmake --build . --target ntoskrnl

Make sure you completely clean your builddir, and re-generate the projects.

File: sdk/cmake/msvc.cmake [hunk 1]
```diff
@@ -536,7 +538,7 @@ function(add_linker_script _target _linker_script_file)
         add_custom_command(
             # OUTPUT ${_generated_file}
             TARGET ${_target} PRE_LINK # PRE_BUILD
-            COMMAND ${CMAKE_C_COMPILER} /nologo ${_no_std_includes_flag} /D__LINKER__ /EP /c "${_file_full_path}" > "${_generated_file}"
+            COMMAND cl /nologo ${_no_std_includes_flag} /D__LINKER__ /EP /c "${_file_full_path}" > "${_generated_file}"
```
Comment (HBelusca) [inline sdk/cmake/msvc.cmake:541]: Why is it required to change the name of the C compiler? Shouldn't this be already correctly configured in `CMAKE_C_COMPILER` for MSVC builds?

**Comment by HBelusca** [comment]
Regarding ASM_MASM, see also https://gitlab.kitware.com/cmake/cmake/-/issues/20738 (opened issue).

File: sdk/cmake/msvc.cmake [hunk 2]
```diff
@@ -17,6 +17,11 @@ elseif(OPTIMIZE STREQUAL "5")
     add_compile_options(/Gy /Ob2 /Os /Ox /GS-)
 endif()
 
+# FIXME: this adds /MT compilation flag for each compilation unit. But if left empty, Visual Studio default to /MD
```
Comment (HBelusca) [inline sdk/cmake/msvc.cmake:20]:
```suggestion
# FIXME: this adds /MT compilation flag for each compilation unit. But if left empty, Visual Studio defaults to /MD
```

File: sdk/cmake/msvc.cmake [hunk 3]
```diff
@@ -187,14 +192,24 @@ if(MSVC_IDE)
     add_definitions(/DLANGUAGE_EN_US)
 else()
     set(CMAKE_RC_COMPILE_OBJECT "<CMAKE_RC_COMPILER> /nologo <INCLUDES> <FLAGS> <DEFINES> ${I18N_DEFS} /fo <OBJECT> <SOURCE>")
+endif()
+
+if (MSVC_IDE)
+    # We use a custom command to preprocess the file. So there is no flag nor defines to account for
```
Comment (HBelusca) [inline sdk/cmake/msvc.cmake:198]:
```suggestion
    # We use a custom command to preprocess the file. So there is no flag nor defines to account for.
```

File: sdk/lib/cpprt/i386/cpprt.s [hunk 4]
```diff
@@ -3,8 +3,8 @@
 .code
 
 MACRO(DEFINE_ALIAS, alias, orig)
-EXTERN &orig : PROC
-ALIAS <&alias> = <&orig>
+EXTERN < &orig > : PROC
+ALIAS < &alias > = < &orig >
```
Comment (HBelusca) [inline sdk/lib/cpprt/i386/cpprt.s:7]: The documentation claims just one pair of brackets are needed, but here somehow you put these (OK) and down, where using the DEFINE_ALIAS, you put again angle brackets (so the macro should resolve with two angle brackets on the left and on the right 🤔 )
Also, why does it cause problems in this file, but not in the following ones ?? :
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/vssapi/msvc.s;hb=91fceab36ee60c5380b5d48706974234b0d49100#l23
https://git.reactos.org/?p=reactos.git;a=blob;f=drivers/ksfilter/ks/msvc.S;hb=91fceab36ee60c5380b5d48706974234b0d49100#l10
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/cpprt/amd64/cpprt.s;hb=91fceab36ee60c5380b5d48706974234b0d49100#l5
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/crt/except/amd64/cpp.s;hb=91fceab36ee60c5380b5d48706974234b0d49100#l32
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/crt/except/i386/cpp.s;hb=91fceab36ee60c5380b5d48706974234b0d49100#l92

File: sdk/cmake/msvc.cmake (hunk 1 above)
Comment (HBelusca) [reply sdk/cmake/msvc.cmake:524]: So ??

**Comment by Extravert-ir** [comment]
```
   Generating shlwapi.def, shlwapi_stubs.c
  Building Custom Rule D:/a/reactos/reactos/src/dll/win32/shlwapi/CMakeLists.txt
  cmake_pch.c
D:\a\reactos\reactos\src\sdk\include\psdk\winternl.h(31,1): fatal error C1189: #error:  "Do not use this header, use NDK!" [D:\a\reactos\reactos\build\dll\win32\shlwapi\shlwapi.vcxproj]
```

File: sdk/cmake/msvc.cmake (hunk 1 above)
Comment (zefklop) [reply sdk/cmake/msvc.cmake]: Done

File: sdk/cmake/msvc.cmake (hunk 2 above)
Comment (zefklop) [reply sdk/cmake/msvc.cmake:20]: Done

File: sdk/lib/cpprt/i386/cpprt.s (hunk 4 above)
Comment (zefklop) [reply sdk/lib/cpprt/i386/cpprt.s]: Fixed. /Cp flag was missing.

File: boot/environ/CMakeLists.txt [hunk 5]
```diff
@@ -85,7 +85,7 @@ if(MSVC)
     if(ARCH STREQUAL "arm")
         add_target_link_flags(bootmgfw "/ignore:4078 /ignore:4254 /DRIVER")
     else()
-        add_target_link_flags(bootmgfw "/ignore:4078 /ignore:4254 /DRIVER /FIXED")
+        add_target_link_flags(bootmgfw "/ignore:4078 /ignore:4254 /DRIVER /DYNAMICBASE:NO /FIXED")
```
Comment (HBelusca) [inline boot/environ/CMakeLists.txt:88]: Why doesn't dynamicbase work when compiling with the IDE only?

File: hal/halx86/CMakeLists.txt [hunk 6]
```diff
@@ -15,30 +15,12 @@ function(add_hal _halname)
...
-    endif()
-
-    if(${_halname} STREQUAL "halpc98")
-        target_compile_definitions(lib_hal_pc98 PRIVATE SARCH_PC98)
-    endif()
-
-    if(${_halname} STREQUAL "halxbox")
-        target_compile_definitions(lib_hal_xbox PRIVATE SARCH_XBOX)
-    endif()
+    foreach(_component ${_haldata_COMPONENTS})
```
Comment (HBelusca) [inline hal/halx86/CMakeLists.txt:18]: And what happens now to the usage of the specific compile-time definitions necessary for these different versions of the HAL?

Comment (HBelusca) [reply hal/halx86/CMakeLists.txt:18]: EDIT: Ok it may appear that putting these in their corresponding cmake files do the trick indeed....

Comment (zefklop) [reply hal/halx86/CMakeLists.txt:18]: it sure does. I don't get why they where in this function in the first place.

File: boot/environ/CMakeLists.txt (hunk 5 above)
Comment (zefklop) [reply boot/environ/CMakeLists.txt:88]: Visual uses yes as a default value, I guess. Same as with this /MD stuff.

File: toolchain-msvc.cmake [hunk 7]
```diff
@@ -53,13 +53,14 @@ endif()
...
 elseif(ARCH STREQUAL "arm")
-    set(CMAKE_ASM_COMPILER armasm)
+    set(CMAKE_ASM_MASM_COMPILER armasm)
 elseif(ARCH STREQUAL "arm64")
-    set(CMAKE_ASM_COMPILER armasm64)
+    set(CMAKE_ASM_MASM_COMPILER armasm64)
 else()
-    set(CMAKE_ASM_COMPILER ml)
+    set(CMAKE_ASM_MASM_COMPILER ml)
+    set(CMAKE_ASM_MASM_FLAGS_INIT "/Cp")
```
Comment (HBelusca) [inline toolchain-msvc.cmake:63]: Doesn't ML64 also need this compile define?

Comment (zefklop) [reply toolchain-msvc.cmake:64]: Sure. Done.

File: sdk/cmake/msvc.cmake [hunk 8]
```diff
@@ -442,37 +446,26 @@ function(allow_warnings __module)
...
-                DEPENDS ${_source_file_full_path})
-            set_source_files_properties(${_object_file} PROPERTIES EXTERNAL_OBJECT TRUE)
-            list(APPEND ${_target} ${_object_file})
         endforeach()
-    else()
-        list(APPEND ${_target} ${ARGN})
-    endif()
+        add_custom_command(
+            OUTPUT ${_preprocessed_asm_file}
+            COMMAND cl /nologo /X /I${REACTOS_SOURCE_DIR}/sdk/include/asm /I${REACTOS_BINARY_DIR}/sdk/include/asm ${_directory_includes} ${_source_file_defines} ${_directory_defines} /D__ASM__ /D_USE_ML /EP /c ${_source_file_full_path} > ${_preprocessed_asm_file}
```
Comment (Extravert-ir) [inline sdk/cmake/msvc.cmake:465]: `CMAKE_C_COMPILER`?

File: boot/freeldr/freeldr/CMakeLists.txt [hunk 9]
```diff
@@ -282,9 +282,9 @@ if(MSVC)
         add_target_link_flags(freeldr_pe "/ignore:4078 /ignore:4254 /DRIVER")
         add_target_link_flags(freeldr_pe_dbg "/ignore:4078 /ignore:4254 /DRIVER")
     else()
-        add_target_link_flags(freeldr_pe "/ignore:4078 /ignore:4254 /DRIVER /FIXED /FILEALIGN:0x200 /ALIGN:0x200")
+        add_target_link_flags(freeldr_pe "/ignore:4078 /ignore:4254 /DRIVER /FIXED /FILEALIGN:512 /ALIGN:512")
         add_linker_script(freeldr_pe freeldr_i386.msvc.lds)
-        add_target_link_flags(freeldr_pe_dbg "/ignore:4078 /ignore:4254 /DRIVER /FIXED /FILEALIGN:0x200 /ALIGN:0x200")
+        add_target_link_flags(freeldr_pe_dbg "/ignore:4078 /ignore:4254 /DRIVER /FIXED /FILEALIGN:512 /ALIGN:512")
```
Comment (HBelusca) [inline boot/freeldr/freeldr/CMakeLists.txt:287]: Is that a regression of their linker?

Comment (zefklop) [reply boot/freeldr/freeldr/CMakeLists.txt:287]: No, it's MSBUILD complaining, it must be extracting the value from the flag:
`2021-02-03T15:13:57.1115193Z C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(1084,9): error MSB4030: "0x200" is an invalid value for the "SectionAlignment" parameter of the "Link" task. The "SectionAlignment" parameter is of type "System.Int32". [D:\a\reactos\reactos\build\boot\freeldr\freeldr\freeldr_pe.vcxproj]`

Comment (HBelusca) [reply boot/freeldr/freeldr/CMakeLists.txt:287]: lmfao 🤣 🤣

**Comment by Extravert-ir** [comment]
It builds for me locally now, don't know why it's shows errors here

**Review by learn-more — APPROVED** [review]
Builds again for me, haven't checked the changes itself.

**Comment by Extravert-ir** [comment]
Yay!
Screenshot_20210303_070306

**Pull request merged**